### DISCUSSION TITLE
Snippet to help with installing the module

### DISCUSSION THREAD

**Discussion by SamErde** [microsoftgraph/entra-powershell#368] [General]
I made this slightly expanded installation snippet to help in new test environments that might not have all of the pre-requisite modules installed yet:  

```powershell
$RequiredModules = (@'
Microsoft.Graph.DirectoryObjects
Microsoft.Graph.Users
Microsoft.Graph.Users.Actions
Microsoft.Graph.Users.Functions
Microsoft.Graph.Groups
Microsoft.Graph.Identity.DirectoryManagement
Microsoft.Graph.Identity.Governance
Microsoft.Graph.Identity.SignIns
Microsoft.Graph.Applications
Microsoft.Graph.Beta.Applications
Microsoft.Graph.Beta.Users
Microsoft.Graph.Beta.Users.Actions
Microsoft.Graph.Beta.Users.Functions
Microsoft.Graph.Beta.Groups
Microsoft.Graph.Beta.Identity.DirectoryManagement
Microsoft.Graph.Beta.Identity.Governance
Microsoft.Graph.Beta.Identity.SignIns
Microsoft.Graph.Beta.Reports
'@).Split("`n")

# Check if the pre-requisite modules are installed and install them if needed
foreach ($module in $RequiredModules) {
    Write-Host -ForegroundColor Yellow -BackgroundColor DarkBlue "Checking for $module"
    if (!(Get-Module -Name $module -ListAvailable)) {
        Install-Module -Name $module -Scope CurrentUser
    }
}

# Unblock the extracted module files in the current location
Get-ChildItem -Recurse | Unblock-File

# Temporarily add the current folder to your module's path. To make it permanent, run the following with admin privileges:
$env:PSModulePath += ";$((Get-Location).Path)"

# Import the module
Import-Module Microsoft.Graph.Entra -Force
```

**Comment by alexandair** [comment]
`Get-Module -Name $module` will tell you if the module is loaded in the current session, not if the module is installed.

What you need is `Get-Module -Name $module -ListAvailable`.

**Reply by SamErde** [reply]
Can't believe I forgot that bit.
Thanks! 
Fixed.

**Comment by SteveMutungi254** [comment]
Bumping this up: @adamedx, @rtintor, @v-varshamane, @FehintolaObafemi.

**Comment by SamErde** [comment]
Note: this isnt what I'd consider production ready, as there are no checks regarding which files are being unblocked, etc. Could be a starting point for someone who has already downloaded the module or for further development if useful.

**Comment by SamErde** [comment]
Adding beta modules per #377 and found in `RequiredModules` in the package's PSD1 file:

- Microsoft.Graph.Beta.Applications
- Microsoft.Graph.Beta.Users
- Microsoft.Graph.Beta.Users.Actions
- Microsoft.Graph.Beta.Users.Functions
- Microsoft.Graph.Beta.Groups
- Microsoft.Graph.Beta.Identity.DirectoryManagement
- Microsoft.Graph.Beta.Identity.Governance
- Microsoft.Graph.Beta.Identity.SignIns
- Microsoft.Graph.Beta.Reports

**Comment by DanielBradley1** [comment]
@SamErde @SteveMutungi254 

Not sure if this helps, but this is what I have done:
```powershell
#Import module manifest
$data = Import-PowerShellDataFile "$((Get-installedModule Microsoft.Graph.Entra).additionalMetadata.InstalledLocation)\Microsoft.Graph.Entra.psd1"

#Get required modules
$requiredModules = $data.RequiredModules.ModuleName
  
#Install dependent modules
if ($requiredModules) {
    foreach ($module in $requiredModules) {
        if (!(Get-Module -Name $module -ListAvailable)) {
            Install-Module -Name $module -Scope CurrentUser
        }
    }
}
```

**Reply by SamErde** [reply]
Funny that you posted this today. I was just testing another module over the past weekend and was thinking about how to do it this way. Thanks for answering the question before I even asked it!

**Reply by DanielBradley1** [reply]
Looks like the dependencies are now correctly defined from the PSGallery. So another alternative solution is:

```powershell
$requiredModules = (Find-module Microsoft.Graph.Entra -IncludeDependencies -AllowPrerelease) | Where {$_.Name -ne "Microsoft.Graph.Entra"} | Select -ExpandProperty Name

foreach ($module in $requiredModules) {
    if (!(Get-Module -Name $module -ListAvailable)) {
        Install-Module -Name $module -Scope CurrentUser
    }
}
```

**Reply by SamErde** [reply]
Nice! Once this repo is public in the next couple of days, it will be even easier to write the script that I've wanted to make: automatically pull down, extract, and install the latest build of the module. Ironically, having it published to the PowerShell Gallery also makes that completely unnecessary, so...it would just be for fun?! :)
